.github/workflows/rust-prs.yml

```diff
@@ -0,0 +1,459 @@
+          fetch-depth: 0
+
+      - name: Extract the Rust version to use from the `rust-toolchain.toml` file
+        shell: bash
+        run: |
+          rust_version=$(grep "channel" rust-toolchain.toml | cut -d "=" -f 2 | cut -d " " -f 2)
+          echo "RUST_VERSION=$rust_version" >> $GITHUB_ENV
+
+      - name: Install Rust
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ env.RUST_VERSION }}
+          components: clippy, rustfmt
+
+      - name: Install what-rust-changed
+        shell: bash
+        run: cargo install --locked --git https://github.com/grafbase/what-rust-changed.git --branch obmarg/qmtxorwqlrtl
+
+      - name: Run what-rust-changed
+        id: what-rust-changed
+        shell: bash
+        env:
+          WHAT_RUST_CHANGED_CONFIG: .github/what-rust-changed.toml
+        # TODO: Much of this script could be moved into the what-rust-changed code
+        # (or it could have its own action) but don't have time for that right now.
+        run: |
+          set -euo pipefail
+          HEAD_REF="$(git rev-parse HEAD)"
+          BASE_REF="remotes/origin/${GITHUB_BASE_REF:-main}"
+          echo "Head: $HEAD_REF"
+          echo "Base: $BASE_REF"
+          MERGE_BASE=$(git merge-base $BASE_REF $HEAD_REF)
+          echo "Merge Base: $MERGE_BASE"
+          git checkout $MERGE_BASE
+          cargo metadata > /tmp/base.metadata.json
+          git checkout $HEAD_REF
+          cargo metadata > /tmp/target.metadata.json
+          CHANGED_FILES=$(git diff --no-commit-id --name-only -r $MERGE_BASE HEAD)
+          CHANGES=$(echo $CHANGED_FILES | xargs what-rust-changed /tmp/base.metadata.json /tmp/target.metadata.json)
+          echo "rust=$CHANGES" >> "$GITHUB_OUTPUT"
+          echo "Done.  Output Contents:"
+          echo ""
+          cat $GITHUB_OUTPUT
+
+      - name: Check paths changed
+        uses: dorny/paths-filter@v3
+        id: paths-changed
+        with:
+          filters: |
+            grafbase-docker:
+              - cli/Dockerfile
+            gateway-docker:
+              - gateway/Dockerfile
+
+  check-format:
+    needs: [what-changed]
+    if: fromJson(needs.what-changed.outputs.rust).changed-packages
+    runs-on: ubicloud-standard-2
+    steps:
+      - name: Get sources
+        uses: actions/checkout@v4
+
+      - name: Extract the Rust version to use from the `rust-toolchain.toml` file
+        shell: bash
+        run: |
+          rust_version=$(grep "channel" rust-toolchain.toml | cut -d "=" -f 2 | cut -d " " -f 2)
+          echo "RUST_VERSION=$rust_version" >> $GITHUB_ENV
+
+      - name: Install Rust
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ env.RUST_VERSION }}
+          components: rustfmt
+
+      - name: cargo fmt
+        shell: bash
+        run: |
+          cargo fmt --check
+
+  example-component:
+    if: contains(fromJson(needs.what-changed.outputs.rust).changed-packages, 'integration-tests')
```

Comment:
Also `gateway-binary` and `wasi-component-loader`... We should have a simpler place to change this so we remember when more crates start depending on these examples.